Stickyjs improvements

vovayatsyuk opened the pull request:
1. Added `_sticky` class for "stuck" element 
2. Added spacingTop feature (https://github.com/magento/magento2/pull/6536)
3. Added stickAfter feature, that allows postponing sticking, until element will go out from the screen for `stickAfter` pixels.

**Screenshots**

Default Settings

![sticky-default](https://cloud.githubusercontent.com/assets/306080/19619511/05609b02-9872-11e6-9e50-33c79b8db788.gif)

With spacingTop = 25

![sticky-spacing-top](https://cloud.githubusercontent.com/assets/306080/19619518/0e587856-9872-11e6-8a7a-f586ea11a1c7.gif)

With spacingTop = 25 and stickAfter = 60

![sticky-spacing-top-and-stick-after](https://cloud.githubusercontent.com/assets/306080/19619519/19404b9a-9872-11e6-91ad-7dd53d72b757.gif)

vovayatsyuk committed:
SpacingTop feature added

vovayatsyuk committed:
Trailing spaces was removed

vovayatsyuk committed:
Added stickyClass (css class) feature.

vovayatsyuk committed:
Added offsetTop feature

vasiliyseleznev commented:
Hi @vovayatsyuk,

Thanks for second loop with this improvement. We will go ahead and accept it.
Internal ticket: MAGETWO-60046

okorshenko commented:
@vovayatsyuk could you please resolve conflicts so that we can proceed with PR acceptance?

vovayatsyuk committed:
Merge branch 'develop' into improvements-stickyjs

vovayatsyuk commented:
Done.

okorshenko commented:
@vovayatsyuk thank you for resolving the merge conflicts

okorshenko commented on lib/web/mage/sticky.js:
```diff
@@ -41,8 +44,34 @@ define([
 
             if (!isStatic && this.element.is(':visible')) {
                 offset = $(document).scrollTop() - this.parentOffset;
```
why do we need this line of code? Variable offset will be redefined in any case in line 49 or 51. 
Could you please review this logic?

vovayatsyuk replied:
Maybe I'm missing the point, but it's not redefined at those lines, it just modified with the possible `spacingTop` parameter.

vovayatsyuk replied:
Sorry, can you comment on this? I don't understand how you'd like to modify the source.

okorshenko replied:
looks good. Thank you

zetlen commented on lib/web/mage/sticky.js:
```diff
@@ -41,8 +44,34 @@ define([
             if (!isStatic && this.element.is(':visible')) {
                 offset = $(document).scrollTop() - this.parentOffset;
+
+                if (typeof this.options.spacingTop === 'function') {
```
You should check to see whether `spacingTop` is a Number, or whether it returns a Number.
```js
var spacingTop = this.options.spacingTop || 0;
if (typeof this.options.spacingTop === 'function') {
  spacingTop = this.options.spacingTop();
}
spacingTop = Number(spacingTop);
if (isNaN(spacingTop)) {
  throw new Error('sticky: spacingTop must be a Number');
}
offset += this.options.spacingTop;
```
See below for more notes.

zetlen commented on lib/web/mage/sticky.js:
```diff
@@ -41,8 +44,34 @@ define([
+                        offsetTop = this.options.offsetTop();
+                    } else {
+                        offsetTop = this.options.offsetTop;
+                    }
```
This pattern happens twice, and there may be further configurable numbers, so it probably should be separated into a private function.
At the top of this file:
```js
function optionToNumber(self, option) {
  var value = self.options[option] || 0;
  if (typeof value === 'function') {
    value = self.options[option]();
  }
  var converted = Number(value);
  if (isNaN(converted)) {
    throw Error('sticky: Could not convert supplied option "' + option + '" to Number from "' + value + '"');
  }
  return converted;
}
```

And then here, you can just have:
```js
var offsetTop = optionToNumber(this, 'offsetTop');
if (offset < offsetTop) {
  offset = 0;
}
```

And above, you can just have:
```js
offset += optionToNumber(this, 'spacingTop');
```

vovayatsyuk replied:
Nice optimization, but is it ok to send `this`? Maybe it's better to place the function inside the widget to get rid of `this` parameter? I know the function is not needed there, but I'm not sure what is better: "send this" or create a function inside the widget?

zetlen replied:
It's okay to send `this` from an object method to a locally declared function. It's a way to do real privacy in JavaScript. I see it in lots of codebases! But the way that you implemented it, as an underscore-prefixed method, is also just fine.

zetlen commented on lib/web/mage/sticky.js:
```diff
@@ -41,8 +44,34 @@ define([
+                        offsetTop = this.options.offsetTop;
+                    }
+
+                    if (offset < this.options.offsetTop) {
```
This line is a bug; `this.options.offsetTop` may still be a function. You might have means `if (offset < offsetTop)`. The refactor I suggest in my above comments should fix it, though.

zetlen commented on lib/web/mage/sticky.js:
```diff
@@ -41,8 +44,34 @@ define([
+
+                if (offset &&
+                    this.options.offsetTop &&
+                    !this.element.hasClass(this.options.stickyClass)) {
```
Can you explain the desired logic here? It's hard to tell from the code what `!this.element.hasClass(this.options.stickyClass)` means for the business logic. Does it mean "the element is currently displayed as sticky" or "the element is NOT currently displayed as sticky"? For more readability, you could assign it to a variable `stuck`.

vovayatsyuk replied:
The logic inside of this `if` should work for the non-sticked element, when `offsetTop` is used. (It makes smooth return of the items with `offsetTop` parameter, when scrolling up)

In other words, I check if an element should be stuck (offset > 0), `offsetTop` is used and an element is not yet stuck.

zetlen reviewed (changes requested):
There are some readability issues and at least one bug, but mostly this is great!

vovayatsyuk committed:
Merge branch 'develop' into improvements-stickyjs

vovayatsyuk committed:
Fixed offsetTop comparing bug

vovayatsyuk committed:
Added stuck variable for better 'if' statement readability

vovayatsyuk committed:
Duplicated option value detection logic moved to separate function

zetlen commented:
One last question, @vovayatsyuk : The default "sticky" class is also prepended with an underscore. I don't normally see CSS classes prepended with an underscore. Other widgets in `lib/web/mage` don't do this. Do you have a reason for the underscore? I think it's safe to remove, and not to worry about class collisions in this case.

vovayatsyuk commented:
It was a proposal from [previous PR](https://github.com/magento/magento2/pull/6536#issuecomment-254503987)

An underscored class name is added also to the modal popup when it's [shown](https://github.com/magento/magento2/blob/develop/lib/web/css/source/components/_modals.less#L52).

However, I've just looked at other styles and [didn't find](https://github.com/magento/magento2/blob/develop/lib/web/css/source/lib/_navigation.less#L185) underscore usage there. Looks like there is no convention in CSS class naming...

zetlen commented:
@vovayatsyuk You're right; there appears to be no convention. Until we establish a convention, this is no reason to hold up a PR merge. Thank you for your contribution!

zetlen reviewed (approved):
Approved.

K7Triton commented:
@vovayatsyuk Hi. I can`t apply your improvement.
I did checkout your branch. And I do not have any changes in my application. Behavior is still old.

vovayatsyuk commented:
You should redeploy [static content](http://devdocs.magento.com/guides/v2.1/config-guide/cli/config-cli-subcommands-static-view.html):

Quick solution:

1. Remove old file:

    ```
    $ bin/magento deploy:mode:set developer
    $ find pub/static -regex ".*/sticky.js" -exec rm -rf {} \;;
    ```

2. Refresh the page in browser

K7Triton commented:
I do it same.
- I try install magento from your branch https://github.com/vovayatsyuk/magento2/tree/improvements-stickyjs
- Make composer install
- bin/magento deploy:mode:set developer
- bin/magento static-content:deploy
- but in pub/static (sticky.js is missing)

I Also try apply your commit like a patch for my project and nothing changing.

vovayatsyuk commented:
Sorry, it's a bit off topic to talk about magento deployment here.

If you'd like to see how the patch is working, you can open deployed file and apply the changes manually.

If you use English locale and luma theme, you'll find the file right here: `pub/static/frontend/Magento/luma/en_US/mage/sticky.js`

If that will not work for you, I really don't know how to help you, sorry.

p.s. did you flush browser cache?
p.p.s What changes do you want to see? My changes do not affect any visible parts of magento. It's up to third-party developers to use the extended functionality.

omiroshnichenko commented on lib/web/mage/sticky.js:
```diff
@@ -11,7 +11,26 @@ define([
+            stickyClass: '_sticky'
+        },
+
+        _optionToNumber(option) {
```
This is not correct function declaration. Move it out of widget in to module private scope or declare this like method of widget. Also, i have a question, why are you check on type of function `offsetTop` and `spacingTop` when it's declared below as Number?

zetlen replied:
Hey, thank you @omiroshnichenko, I missed that. ES6 object shorthand is not currently allowed in Magento JavaScript. This should change to:
```js
_optionToNumber: function(option) {
```

omiroshnichenko reviewed (commented):
Please, validate it with static test locally`grunt static --file=lib/web/mage/sticky.js` or rerun travis build.

omiroshnichenko commented on lib/web/mage/sticky.js:
```diff
@@ -11,7 +11,26 @@ define([
+
+            var converted = Number(value);
+            if (isNaN(converted)) {
+                throw Error(
```
We don't have practice to throwing errors in magento js.

omiroshnichenko commented on lib/web/mage/sticky.js:
```diff
@@ -11,7 +11,26 @@ define([
         options: {
-            container: ''
+            container: '',
+            spacingTop: 0,
```
You have been provide new settings properties that by default is Number, no reason to check it on another types. User can't change this property, only developer, when developer will be configuring sticky widget he can look at type of parameter. Also, `offsetTop` it's native property of HTMLElement would be nice to change it to another, e.g. `blockOffsetTop`.

vovayatsyuk replied:
Still thinking about a new name for it :tired_face:

vovayatsyuk replied:
What would you say to rename it to `stickAfter` or `triggerOffset` or `stickOffset`?

vovayatsyuk replied:
I've picked a `stickAfter`. It seems that it better and self-explanatory name.

omiroshnichenko commented on lib/web/mage/sticky.js:
```diff
@@ -40,9 +59,23 @@ define([
+                var stuck = this.element.hasClass(this.options.stickyClass);
+                if (offset && this.options.offsetTop && !stuck) {
+                    var offsetTop = this._optionToNumber('offsetTop');
+                    if (offset < offsetTop) {
```
This check can be include in upper `if` statement via AND(&&) operator.

okorshenko commented:
Hi @vovayatsyuk 
Could you please check review comments?

vovayatsyuk committed:
Shorthand method definition syntax replaced with standard one

vovayatsyuk committed:
OptionToNumber method refactored:

- Removed option type validation
- Renamed into getOptionValue as it can be used for any option types

vovayatsyuk committed:
Better code readability

vovayatsyuk committed:
Fixed JSHint notices

vovayatsyuk committed:
ESLint and JSCS fixes

vovayatsyuk committed:
offsetTop option renamed into stickAfter as it has nothing with element's offsetTop property

vovayatsyuk commented:
Sorry for the delay. I think I'm done.

omiroshnichenko commented on lib/web/mage/sticky.js:
```diff
@@ -11,7 +11,26 @@ define([
+        _getOptionValue: function (option) {
+            var value = this.options[option] || 0;
+
+            if (typeof value === 'function') {
```
Why you check value on function type? In what case it can be function?

vovayatsyuk replied:
It could be useful to set a function for `stickAfter` parameter that will calculate sticky element height when you wish to stick element after it goes out of the viewport and the element could change its height dynamically (shopping cart for example)
You can use different spacing top value for different screen size/orientation also.

omiroshnichenko replied:
Ok, but can you leave docblock with description. Thanks.

vovayatsyuk replied:
Done.

vovayatsyuk committed:
DocBlocks added for the new options

vovayatsyuk committed:
DocBlock added for the 'container' option

vovayatsyuk committed:
Typo error in DocBlock fixed
